File: src/ISA/ISA/JsonTypes/ProcessParameterValue.fs
namespace ISA


type ProcessParameterValue =
    {
        Category    : ProtocolParameter option
        Value       : Value option
        Unit        : OntologyAnnotation option
    }

    static member make category value unit : ProcessParameterValue = 
        {
            Category = category
            Value = value
            Unit = unit
        }

    static member create (?Category,?Value,?Unit) : ProcessParameterValue = 
        ProcessParameterValue.make Category Value Unit

    static member empty =
        ProcessParameterValue.create()

    /// Returns the name of the category as string
    member this.NameText =
        this.Category
        |> Option.map (fun oa -> oa.NameText)
        |> Option.defaultValue ""

    member this.ValueText =
    
        this.Value
        |> Option.map (fun oa ->
            match oa with
            | Value.Ontology oa  -> oa.NameText
            | Value.Float f -> string f
            | Value.Int i   -> string i
            | Value.Name s  -> s
        )
        |> Option.defaultValue ""

    member this.ValueWithUnitText =
        let unit = 
            this.Unit |> Option.map (fun oa -> oa.NameText)
        let v = this.ValueText
        match unit with
        | Some u    -> sprintf "%s %s" v u
        | None      -> v

    member this.MapCategory(f : OntologyAnnotation -> OntologyAnnotation) =
        {this with Category = this.Category |> Option.map (fun p -> p.MapCategory f) }

    member this.SetCategory(c : OntologyAnnotation) =
        {this with Category = 
                            match this.Category with
                            | Some p -> Some (p.SetCategory c)
                            | None -> Some (ProtocolParameter.create(ParameterName = c))
        }

    interface IISAPrintable with
        member this.Print() =
            this.ToString()
        member this.PrintCompact() =
            let category = this.Category |> Option.map (fun f -> f.NameText)
            let unit = this.Unit |> Option.map (fun oa -> oa.NameText)
            let value = 
                this.Value
                |> Option.map (fun v ->
                    let s = (v :> IISAPrintable).PrintCompact()
                    match unit with
                    | Some u -> s + " " + u
                    | None -> s
                )
            match category,value with
            | Some category, Some value -> category + ":" + value
            | Some category, None -> category + ":" + "No Value"
            | None, Some value -> value
            | None, None -> ""

File: js/ISA/ISA/JsonTypes/ProcessParameterValue.js
import { Record, toString } from "../../../fable_modules/fable-library.4.1.4/Types.js";
import { bind, defaultArg, map } from "../../../fable_modules/fable-library.4.1.4/Option.js";
import { ProtocolParameter__SetCategory_Z4C0FE73C, ProtocolParameter_create_Z6C54B221, ProtocolParameter__MapCategory_Z69DD836A, ProtocolParameter__get_TryNameText, ProtocolParameter_$reflection, ProtocolParameter__get_NameText } from "./ProtocolParameter.js";
import { OntologyAnnotation_$reflection, OntologyAnnotation__get_NameText } from "./OntologyAnnotation.js";
import { record_type, option_type } from "../../../fable_modules/fable-library.4.1.4/Reflection.js";
import { Value_$reflection } from "./Value.js";
import { int32ToString } from "../../../fable_modules/fable-library.4.1.4/Util.js";
import { printf, toText } from "../../../fable_modules/fable-library.4.1.4/String.js";

export class ProcessParameterValue extends Record {
    constructor(Category, Value, Unit) {
        super();
        this.Category = Category;
        this.Value = Value;
        this.Unit = Unit;
    }
    Print() {
        const this$ = this;
        return toString(this$);
    }
    PrintCompact() {
        let value_2, category_2, category_1, value_1;
        const this$ = this;
        const category = map(ProtocolParameter__get_NameText, this$.Category);
        const unit = map(OntologyAnnotation__get_NameText, this$.Unit);
        const value = map((v) => {
            const s = v.PrintCompact();
            if (unit == null) {
                return s;
            }
            else {
                return (s + " ") + unit;
            }
        }, this$.Value);
        return (category == null) ? ((value == null) ? "" : ((value_2 = value, value_2))) : ((value == null) ? ((category_2 = category, (category_2 + ":") + "No Value")) : ((category_1 = category, (value_1 = value, (category_1 + ":") + value_1))));
    }
}

export function ProcessParameterValue_$reflection() {
    return record_type("ARCtrl.ISA.ProcessParameterValue", [], ProcessParameterValue, () => [["Category", option_type(ProtocolParameter_$reflection())], ["Value", option_type(Value_$reflection())], ["Unit", option_type(OntologyAnnotation_$reflection())]]);
}

export function ProcessParameterValue_make(category, value, unit) {
    return new ProcessParameterValue(category, value, unit);
}

export function ProcessParameterValue_create_569825F3(Category, Value, Unit) {
    return ProcessParameterValue_make(Category, Value, Unit);
}

export function ProcessParameterValue_get_empty() {
    return ProcessParameterValue_create_569825F3();
}

/**
 * Returns the name of the category as string
 */
export function ProcessParameterValue__get_NameText(this$) {
    return defaultArg(map(ProtocolParameter__get_NameText, this$.Category), "");
}

/**
 * Returns the name of the category as string
 */
export function ProcessParameterValue__get_TryNameText(this$) {
    return bind(ProtocolParameter__get_TryNameText, this$.Category);
}

export function ProcessParameterValue__get_ValueText(this$) {
    return defaultArg(map((oa) => {
        switch (oa.tag) {
            case 2:
                return oa.fields[0].toString();
            case 1:
                return int32ToString(oa.fields[0]);
            case 3:
                return oa.fields[0];
            default:
                return OntologyAnnotation__get_NameText(oa.fields[0]);
        }
    }, this$.Value), "");
}

export function ProcessParameterValue__get_ValueWithUnitText(this$) {
    const unit = map(OntologyAnnotation__get_NameText, this$.Unit);
    const v = ProcessParameterValue__get_ValueText(this$);
    if (unit == null) {
        return v;
    }
    else {
        const u = unit;
        return toText(printf("%s %s"))(v)(u);
    }
}

export function ProcessParameterValue__MapCategory_Z69DD836A(this$, f) {
    return new ProcessParameterValue(map((p) => ProtocolParameter__MapCategory_Z69DD836A(p, f), this$.Category), this$.Value, this$.Unit);
}

export function ProcessParameterValue__SetCategory_Z4C0FE73C(this$, c) {
    let matchValue;
    return new ProcessParameterValue((matchValue = this$.Category, (matchValue == null) ? ProtocolParameter_create_Z6C54B221(void 0, c) : ProtocolParameter__SetCategory_Z4C0FE73C(matchValue, c)), this$.Value, this$.Unit);
}

/**
 * Returns the name of the paramater value as string if it exists
 */
export function ProcessParameterValue_tryGetNameText_5FD7232D(pv) {
    return ProcessParameterValue__get_TryNameText(pv);
}

/**
 * Returns the name of the paramater value as string
 */
export function ProcessParameterValue_getNameText_5FD7232D(pv) {
    return ProcessParameterValue__get_NameText(pv);
}

/**
 * Returns true if the given name matches the name of the parameter value
 */
export function ProcessParameterValue_nameEqualsString(name, pv) {
    return ProcessParameterValue__get_NameText(pv) === name;
}

export function ProcessParameterValue_getCategory_5FD7232D(pv) {
    return pv.Category;
}

